modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -3814,4 +3845,22 @@ private HybridTimestamp indexStartBuildingActivationTs(BuildIndexReplicaRequest
 
         return hybridTimestamp(catalog.time());
     }
+
+    private int schemaVersionByTs(HybridTimestamp ts) {
+        int activeCatalogVersion = catalogService.activeCatalogVersion(ts.longValue());
+
+        CatalogTableDescriptor table = catalogService.table(tableId(), activeCatalogVersion);
+
+        assert table != null : "tableId=" + tableId() + ", catalogVersion=" + activeCatalogVersion;
+
+        return table.tableVersion();
+    }
+
+    private static @Nullable BinaryRow binaryRow(@Nullable TimedBinaryRow timedBinaryRow) {
+        return timedBinaryRow == null ? null : timedBinaryRow.binaryRow();
+    }
+
+    private @Nullable BinaryRow upgrage(@Nullable BinaryRow source, int targetSchemaVersion) {
+        return source == null ? null : new BinaryRowUpgrader(schemaRegistry, targetSchemaVersion).upgrade(source);
```

Comment:
I wonder if this creates too much garbage. For each table version, we only need one upgrader, it can be cached.

Reply:
We discussed it in person, let's leave it for now.